resolver/projects.go

```diff
@@ -38,6 +39,28 @@ func LocateGoModProjects(projects []GoModProject) (map[string]Dependency, error)
 	return deps, nil
 }
 
+func LocateNpmPackageLockV3Projects(root string, projects []Project) (map[string]Dependency, error) {
+	nodeModulePrefixRegexp := regexp.MustCompile(`.*node_modules\/`)
+	deps := make(map[string]Dependency, len(projects))
+
+	for _, project := range projects {
+		baseDependencyName := nodeModulePrefixRegexp.ReplaceAllString(project.Name(), "")
+		sourceDir := filepath.Join(root, project.Name())
+		if _, err := os.Stat(sourceDir); os.IsNotExist(err) {
+			// If a direct path to a nested dependency is not found, it was hoisted to the root.
+			sourceDir = filepath.Join(root, baseDependencyName)
```

Comment:
can you give an example? I'm not familiar enough with npm to know what this means

Reply:
Example with stackrox/stackrox today:

In the `package-lock.json` file we have a dependency `type-is` that itself has a dependency on `mime-types`. Since this is a transitive dependency, our package-lock will list the naive path of `"node_modules/type-is/node_modules/mime-types"` as the path to the `mime-types` dependency on disk. (npm dependencies often have multiple levels of `node_modules` under direct dependencies in the project's top level `node_modules` to handle transitive dependencies that span multiple versions -- in order to avoid conflicts.)

In our project right now, `type-is` depends on `mime-types` version `~2.1.24`, i.e. any patch version `2.1.x` where `x` is >= 24. However during the actual module installation for our project, `npm` determined that `mime-types` version `2.1.35` could satisfy the dependency requirements of `type-is` and all other modules in the project and hoisted it to the top level directly under `node_modules/mime-types`. This is to save disk space, download time, and build size.

So even though our package lock lists the dependency path as `"node_modules/type-is/node_modules/mime-types"`, at module installation time and build time the true path to this module is just `"node_modules/mime-types"`, which is what this line of code handles. Note that this isn't always the case, as sometimes due to dependency version conflicts the module actually does reside at the nested path.

I'll update the comment here to explain similarly to the example above.